### PR TITLE
Port monitor-level filtering functionality from signalfx-agent

### PR DESCRIPTION
**Pull request by dmitryax**
The following files ported mostly as is from signalfx-agent:
- internal/receiver/smartagentreceiver/filtering.go 
- internal/receiver/smartagentreceiver/filtering.go

### CODE REVIEW

File: internal/receiver/smartagentreceiver/filtering.go [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+// HasEnabledMetricInGroup returns true if there are any metrics enabled that
+// fall into the given group.
+func (mf *monitorFiltering) HasEnabledMetricInGroup(group string) bool {
+	if mf.metadata == nil {
+		return false
+	}
+
+	for _, m := range mf.EnabledMetrics() {
+		// TODO: If metric names in metadata.yaml ever support wildcards this
```
Comment (rmfitzpatrick) [inline internal/receiver/smartagentreceiver/filtering.go:82]: Is this TODO still valid for this project?

**Comment by rmfitzpatrick** [comment]
Are you able to add config file tests using `datapointsToExclude`, `extraMetrics`, `extraGroups` and other applicable fields?  Ideally the diff coverage would also be more complete.

File: internal/receiver/smartagentreceiver/filtering_test.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+import (
+	"fmt"
+	"testing"
+
+	"github.com/signalfx/golib/v3/datapoint"
+	"github.com/signalfx/signalfx-agent/pkg/monitors"
+	"github.com/signalfx/signalfx-agent/pkg/utils"
+)
+
+func testMetadata(sendUnknown bool) *monitors.Metadata {
```
Comment (rmfitzpatrick) [inline internal/receiver/smartagentreceiver/filtering_test.go:26]: Mind attributing this test file content as well?

File: internal/receiver/smartagentreceiver/receiver.go [hunk 3]
```diff
@@ -131,7 +131,20 @@ func (r *Receiver) createMonitor(monitorType string, host component.Host) (monit
...
 	monitor = monitorFactory() // monitor is a pointer to a monitor struct
 
-	output := NewOutput(*r.config, r.nextConsumer, host, r.logger)
+	// Make metadata nil if we aren't using built in filtering and then none of
+	// the new filtering logic will apply.
+	metadata, ok := monitors.MonitorMetadatas[monitorType]
+	if !ok || metadata == nil {
+		// This indicates a programming error in not specifying metadata, not
+		// bad user input
+		panic(fmt.Sprintf("could not find monitor metadata of type %s", monitorType))
```
Comment (rmfitzpatrick) [inline internal/receiver/smartagentreceiver/receiver.go:140]: The current pattern is to return errors and not panic.  Looking through collector and collector-contrib panicking doesn't seem too common either.

File: internal/receiver/smartagentreceiver/filtering.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+func (mf *monitorFiltering) AddDatapointExclusionFilter(filter dpfilters.DatapointFilter) {
+	mf.filterSet.ExcludeFilters = append(mf.filterSet.ExcludeFilters, filter)
+}
+
+func (mf *monitorFiltering) EnabledMetrics() []string {
+	if mf.metadata == nil {
+		return nil
+	}
+
+	// This can be optimized to cache results on the struct if it is worth it.
```
[the remark this answers is not in the document]
Comment (dmitryax) [reply internal/receiver/smartagentreceiver/filtering.go:60]: Looks like it is not, removing the comment

File: internal/receiver/smartagentreceiver/filtering.go (hunk 1 above)
Comment (dmitryax) [reply internal/receiver/smartagentreceiver/filtering.go:82]: removing

File: internal/receiver/smartagentreceiver/filtering_test.go (hunk 2 above)
Comment (dmitryax) [reply internal/receiver/smartagentreceiver/filtering_test.go:28]: done

File: internal/receiver/smartagentreceiver/receiver.go (hunk 3 above)
Comment (dmitryax) [reply internal/receiver/smartagentreceiver/receiver.go]: converted to a returnen error

**Comment by dmitryax** [comment]
@rmfitzpatrick , I added the config tests and bumped the coverage. PTAL

File: internal/receiver/smartagentreceiver/config_test.go [hunk 5]
```diff
@@ -416,3 +417,77 @@ func TestLoadInvalidConfigWithNonStringArrayDimensionClients(t *testing.T) {
...
+						MetricNames: []string{"./[0-"},
+					},
+				},
+			},
+		},
+	}, fsCfg)
+
+	err = fsCfg.validate()
+	require.Error(t, err)
+	require.EqualError(t, err, "unexpected end of input")
```
Comment (rmfitzpatrick) [inline internal/receiver/smartagentreceiver/config_test.go:492]: Not necessary to address here, but we should probably provide an error wrapper for validation errors so they provide a bit more context.  This is pretty vague.

File: internal/receiver/smartagentreceiver/filtering.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+				return nil
+			}
+		}
+
+		logrus.Warnf("extraMetrics: metric pattern '%s' did not match any available metrics for monitor %s",
+			metricName, metadata.MonitorType)
+	}
+
+	if !metadata.HasMetric(metricName) {
+		logrus.Warnf("extraMetrics: metric '%s' does not exist for monitor %s", metricName, metadata.MonitorType)
```
Comment (rmfitzpatrick) [inline internal/receiver/smartagentreceiver/filtering.go:171]: We should be using the receiver's zap logger and not the global logrus logger.

**Comment by rmfitzpatrick** [comment]
@dmitryax thanks for this and your updates.  It lgtm w/ all logrus usage converted to zap (we have a shim and I think it would end up using the default instance which wouldn't be ideal).

File: internal/receiver/smartagentreceiver/config_test.go (hunk 5 above)
Comment (dmitryax) [reply internal/receiver/smartagentreceiver/config_test.go:492]: Makes sense, submitted an issue https://github.com/signalfx/splunk-otel-collector/issues/157. Will cover it in a following PR

File: internal/receiver/smartagentreceiver/filtering.go (hunk 6 above)
Comment (dmitryax) [reply internal/receiver/smartagentreceiver/filtering.go]: addressed

**Review by rmfitzpatrick — APPROVED** [review]

**Pull request merged**